maintain the given order on step execution

onur-ozkan opened the pull request:
Previously step execution disregarded the CLI order and this change executes the given steps in the order specified on CLI.

For example, running `x $kind a b c` will execute `$kind` step for `a`, then `b`, then `c` crates in the specified order.

Fixes #126165
cc @matthiaskrgr

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9887943227/job/27310772719) [(plain)](https://github.com/rust-lang/rust/commit/4abd67f7cd5ae6e52ffc8e88adb17cabe6cf4cda/checks/27310772719/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 46)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 49)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 61)
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.61s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
   Compiling pretty_assertions v1.4.0
error[E0599]: no method named `intersection_removing_matches` found for enum `core::builder::PathSet` in the current scope
##[error]   --> src/core/builder/tests.rs:130:30
    |
130 |     let subset = library_set.intersection_removing_matches(&mut command_paths, Kind::Build);
    |
   ::: src/core/builder.rs:220:1
    |
220 | pub enum PathSet {
220 | pub enum PathSet {
    | ---------------- method `intersection_removing_matches` not found for this enum
    |
help: there is a method `intersection_pathset` with a similar name
    |
130 |     let subset = library_set.intersection_pathset(&mut command_paths, Kind::Build);

For more information about this error, try `rustc --explain E0599`.
error: could not compile `bootstrap` (lib test) due to 1 previous error
Build completed unsuccessfully in 0:38:24
```

</details>

jieyouxu commented on src/bootstrap/src/core/builder.rs:
```diff
@@ -473,14 +459,39 @@ impl StepDescription {
+            }
+        });
+
+        assert!(steps_to_run.is_empty(), "Some steps did not executed.");
```
Nit: "some steps were not executed".
Discussion: In case of failure, should it also print which steps were not executed?

onur-ozkan commented:
I will re-work on this later again.

@rustbot author

onur-ozkan committed:
maintain the given order on step execution

Previously step execution disregarded the CLI order and this change executes the given
steps in the order specified on CLI.

For example, running `x $kind a b c` will execute `$kind` step for `a`, then `b`, then `c` crates
in the specified order.

Signed-off-by: onur-ozkan <work@onurozkan.dev>

onur-ozkan commented:
Simplified the solution without refactoring the entire internals about path resolution (which was the case initially).

@rustbot ready

Kobzol commented:
I'm a bit confused by the original issue. It used `x test A B` as an example, but most tests should be handled by [this](https://github.com/rust-lang/rust/blob/73028fe48320766d5dd931394c2d7b17bed14646/src/bootstrap/src/core/builder.rs#L464), and thus they should already be handled in the original order. Is there an example of a different command where order is important, but which is not handled by the suite logic?

onur-ozkan commented:
> It used x test A B as an example, but most tests should be handled by [this](https://github.com/rust-lang/rust/blob/73028fe48320766d5dd931394c2d7b17bed14646/src/bootstrap/src/core/builder.rs#L464), and thus they should already be handled in the original order.

That's not quite right. The logic below handles the step executions most of the time. The one you linked is supposed to handle specific test paths (like `tests/something/test.rs`).

To debug the ordering, you can run this command:

`x test clippy miri bootstrap rustdoc sysroot --dry-run | grep Testing`

and compare the CLI order to the executed steps.

Kobzol commented:
Ah, I see, thanks. I guess that I usually run specific test paths, so I didn't think of this. I'll test it locally.

Kobzol commented on src/bootstrap/src/core/builder.rs:
```diff
@@ -473,9 +473,42 @@ impl StepDescription {
+            //
+            // If we resolve the step's path from the given CLI input, this value will be updated with
+            // the step's actual index.
+            let mut closest_index = usize::MAX;
```
The logic is a bit convoluted. What about doing something like this instead?
```rust
for (desc, should_run) in v.iter().zip(&should_runs) {
    let orig_paths = paths.clone();
    let pathsets = should_run.pathset_for_paths_removing_matches(&mut paths, desc.kind);

    // Find the lowest index of a removed path
    let index = orig_paths.iter().position(|p| !paths.contains(p)).unwrap_or(usize::MAX);
    steps_to_run.push((index, desc, pathsets));
}
```
Then we wouldn't need `indexed_paths` or the other logic (unless I'm missing something).

onur-ozkan replied:
This won't work. Cloning updated `paths` in each iteration can lead to use of same indexes multiple times (e.g, we may end up using `0` more than once), so the sorting will result with an incorrect order.

Kobzol replied:
Sorry, I combined too many ideas at once :facepalm: `orig_paths` here was supposed to serve for detecting removed paths, but to get the actual position, we can use another approach:

```rust
let mut steps_to_run = vec![];
let orig_paths = paths.clone();

for (desc, should_run) in v.iter().zip(&should_runs) {
    let previous_paths = paths.clone();
    let pathsets = should_run.pathset_for_paths_removing_matches(&mut paths, desc.kind);

    // Find the first path that was removed and its original index
    let index = previous_paths
        .iter()
        .find(|p| !paths.contains(p))
        .and_then(|p| orig_paths.iter().position(|p2| p2 == p))
        .unwrap_or(usize::MAX);
    steps_to_run.push((index, desc, pathsets));
}
```
If we want to remove the paths after being found (this is probably only useful to handle duplicates in `paths`?), we could use a hash map:
```rust
let mut orig_paths: HashMap<PathBuf, usize> =
  paths.iter().enumerate().map(|(i, p)| (p.clone(), i)).collect();

for (desc, should_run) in v.iter().zip(&should_runs) {
  let previous_paths = paths.clone();
  let pathsets = should_run.pathset_for_paths_removing_matches(&mut paths, desc.kind);

  // Find the first path that was removed and its original index
  let index = previous_paths
      .iter()
      .find(|p| !paths.contains(p))
      .and_then(|p| orig_paths.remove(p))
      .unwrap_or(usize::MAX);
  steps_to_run.push((index, desc, pathsets));
}
```

onur-ozkan replied:
Isn't the current approach more straight-forward? Specially the first examples adds  more overhead while not really helping in terms of readability.

Kobzol replied:
I guess it's a matter of style. To me it was a bit difficult to understand the interaction of the added Vecs. In my version, the discovery of removed paths is local to the body of the for loop, so it's easier for me to understand.

But as I said, this is purely a stylistic thing. If you prefer your version, feel free to r=me :)

onur-ozkan replied:
What about [this](https://github.com/rust-lang/rust/compare/774af29940f790204d9da551ae4f42fe899d3d56..4ad753af915840ca9de07dd2df80419b57136872) approach? Removed the outer vector of indexed paths.

Kobzol replied:
Ok, nice, without the indexed vector it's easier to understand. The hashmap solution would be too complicated in this version, since it would need to track both path usage and the index, and that gets annoying (I tried it).

So LGTM :+1:

Kobzol reviewed (commented):
This is the first time I actually saw dry run being useful :) However it was in the mode where it was enabled manually (`--dry-run`), I'd still get rid of the double execution of all steps :laughing:

onur-ozkan commented:
@bors r=Kobzol

bors commented:
:pushpin: Commit 4ad753af915840ca9de07dd2df80419b57136872 has been approved by `Kobzol`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Kobzol 4ad753af915840ca9de07dd2df80419b57136872 -->
<!-- homu: {"type":"Approved","sha":"4ad753af915840ca9de07dd2df80419b57136872","approver":"Kobzol","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 4ad753af915840ca9de07dd2df80419b57136872 with merge 3de0a7c716daab58bdb6551b0d0af6a466486639...
<!-- homu: {"type":"BuildStarted","head_sha":"4ad753af915840ca9de07dd2df80419b57136872","merge_sha":"3de0a7c716daab58bdb6551b0d0af6a466486639"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9972934607/job/27563695252)
Approved by: Kobzol
Pushing 3de0a7c716daab58bdb6551b0d0af6a466486639 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Kobzol","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/9972934607/job/27563695252"},"merge_sha":"3de0a7c716daab58bdb6551b0d0af6a466486639"} -->

rust-timer commented:
Finished benchmarking commit (3de0a7c716daab58bdb6551b0d0af6a466486639): [comparison URL](https://perf.rust-lang.org/compare.html?start=cb12b52f160c633551bd6e5be8d4c6e2e8660950&end=3de0a7c716daab58bdb6551b0d0af6a466486639&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=cb12b52f160c633551bd6e5be8d4c6e2e8660950&end=3de0a7c716daab58bdb6551b0d0af6a466486639&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=cb12b52f160c633551bd6e5be8d4c6e2e8660950&end=3de0a7c716daab58bdb6551b0d0af6a466486639&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=cb12b52f160c633551bd6e5be8d4c6e2e8660950&end=3de0a7c716daab58bdb6551b0d0af6a466486639&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=cb12b52f160c633551bd6e5be8d4c6e2e8660950&end=3de0a7c716daab58bdb6551b0d0af6a466486639&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 768.259s -> 768.642s (0.05%)
**Artifact size**: 328.63 MiB -> 328.71 MiB (0.02%)
